courses/programming-in-haskell/pih-exercises/src/Chapter08.hs

```diff
@@ -0,0 +1,44 @@
+module Chapter08 (Nat(..), mult, Expr(..), folde, Tree(..), complete) where
+
+data Nat = Zero
+         | Succ Nat deriving (Show, Eq)
+
+add :: Nat -> Nat -> Nat
+add Zero n = n
+add (Succ m) n = Succ(add m n)
+
+mult :: Nat -> Nat -> Nat
+mult Zero n = Zero
+mult (Succ m) n = add (mult m n) n
+
+
+data Expr = Val Int
+          | Add Expr Expr
+          | Mul Expr Expr
+
+folde :: (Int -> a) -> (a -> a-> a) -> (a -> a -> a) -> (Expr -> a)
```

Comment:
As far as I can see the last parenthesis is not needed, i.e.:

    folde :: (Int -> a) -> (a -> a-> a) -> (a -> a -> a) -> Expr -> a